Add -ffunction-sections flag

nurmukhametov opened the pull request:
It forces the compiler to place each function in its own section.

This PR addresses the issue #2902. One can eliminate unused copies of exported functions providing this flag and enabling a linker flag that removes unused sections (`-gc-sections` for `ld` and compatible linkers).

aneshlya commented on src/main.cpp:
```diff
@@ -104,6 +104,7 @@ static void lPrintVersion() {
     printf("        fast\t\t\t\tUse high-performance but lower-accuracy math functions\n");
     printf("        svml\t\t\t\tUse the Intel(r) SVML math libraries\n");
     printf("        system\t\t\t\tUse the system's math library (*may be quite slow*)\n");
+    printf("    [-ffunction-sections]\t\t\tPlace each function in its own section\n");
```
Maybe 
```suggestion
    printf("    [-f[no]function-sections]\t\t\tPlace each function in its own section\n");
```

nurmukhametov replied:
Done

aneshlya reviewed (approved):
Looks good!

dbabokin commented on src/main.cpp:
```diff
@@ -104,6 +104,7 @@ static void lPrintVersion() {
     printf("        fast\t\t\t\tUse high-performance but lower-accuracy math functions\n");
     printf("        svml\t\t\t\tUse the Intel(r) SVML math libraries\n");
     printf("        system\t\t\t\tUse the system's math library (*may be quite slow*)\n");
+    printf("    [-f[no-]function-sections]\t\t\tPlace each function in its own section\n");
```
```suggestion
    printf("    [-f[no-]function-sections]\t\tPlace each function in its own section\n");
```

nurmukhametov replied:
Done

dbabokin reviewed (changes requested):
The changes don't work on macOS - the function is still in TEXT section.

But I tried clang and had the same behavior, i.e. no effect on emitted assembler. Do you know why? It's probably ok to have the same behavior as clang (i.e. accept the option which has no effect), but would be good to understand what's going on.

nurmukhametov commented:
> The changes don't work on macOS - the function is still in TEXT section.
> 
> But I tried clang and had the same behavior, i.e. no effect on emitted assembler. Do you know why? It's probably ok to have the same behavior as clang (i.e. accept the option which has no effect), but would be good to understand what's going on.

Yes. It looks like a specific mechanism is slightly different with macOS toolchain. Although, I didn't manage to find quickly what it is. Anyway, the usage of `-dead_strip` linker option ([docs](https://github.com/llvm/llvm-project/blob/9cd93774098c861c260090a690f428b7ae031c65/clang/include/clang/Basic/AttrDocs.td#L83)) eliminates unused copies regardless of `-ffunction-sections` flag.

dbabokin reviewed (approved):
Could you put this knowledge about macOS (that it doesn't work in the same way on macOS, that symbols are stripped anyway, and that we are following clang, which does have this option on macOS, but it does nothing) to comments in the test?

After that it's good to merge.

nurmukhametov commented:
> Could you put this knowledge about macOS (that it doesn't work in the same way on macOS, that symbols are stripped anyway, and that we are following clang, which does have this option on macOS, but it does nothing) to comments in the test?
> 
> After that it's good to merge.

I have added this comment and updated docs.

dbabokin commented on docs/ispc.rst:
```diff
@@ -852,6 +857,14 @@ position-independent code suitable for dynamic linking avoiding any limit on
+separate section. This flag is useful for reducing the size of the final
+executable by removing unused functions when it is combined with linker flag
+that removes unused sections: ``--gc-sections`` for ``GNU ld`` and ``/OPT:REF``
+for ``MSVC link.exe``. On macOS, this flag does not have any effect (as clang)
```
```suggestion
for ``MSVC link.exe``. On macOS, this flag does not have any effect (as in clang)
```
